Populate runtime with upgrader graph

tugsbayasgalan opened the pull request:
In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69427
* __->__ #68773

Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

tugsbayasgalan committed:
Populate runtime with upgrader graph

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

cccclai commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -0,0 +1,30 @@
+namespace torch {
+namespace jit {
+
+#define IS_IN_POPULATION_PHASE ([] { \
```
neat!

qihqi replied:
Can you explain what you are trying to do?

tugsbayasgalan replied:
So `populate_upgraders_map` inserts upgrader graph entries to an originally empty map. Since we don't want to keep adding stuff to that map everytime torch.jit.load is invoked, we do this to make sure `populate_upgraders_map` is only called once.

qihqi reviewed (commented):
fix the lints

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_upgraders.py:
```diff
@@ -0,0 +1,34 @@
+        torch.jit.load(buffer)
+        upgraders_size_v2 = torch._C.get_upgraders_map_size()
+        self.assertTrue(upgraders_size == upgraders_size_v2)
+        self.assertTrue(upgraders_size > 0)
```
This will break as time goes and we deprecate all upgraders

tugsbayasgalan replied:
Fixed

gmagogsfm commented on test/jit/test_upgraders.py:
```diff
@@ -0,0 +1,34 @@
+        buffer.seek(0)
+        torch.jit.load(buffer)
+        upgraders_size_v2 = torch._C.get_upgraders_map_size()
+        self.assertTrue(upgraders_size == upgraders_size_v2)
```
Maybe check the contents of the upgraders are actually identical between loads?

tugsbayasgalan replied:
Fixed

gmagogsfm commented on torch/jit/_serialization.py:
```diff
@@ -146,6 +146,12 @@ def load(f, map_location=None, _extra_files=None):
+
+    # Populate upgrader map in torch.jit.load because
+    # this is the only place we actually use upgraders
+    from torch.jit.operator_upgraders import populate_upgraders_map
```
Nit:
Move it to top

tugsbayasgalan replied:
It creates circular import issue when on the top.

gmagogsfm replied:
Oh? Could you provide more info? If unconditional import causes circular import issue, then that shows our code structure could use some optimization.

gmagogsfm replied:
Ping on this.

gmagogsfm replied:
Ping.

pmeier replied:
With this import `torch.jit.load` now needs `pyyaml` installed. Is this intended?

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

qihqi commented on torch/csrc/jit/operator_upgraders/upgraders.cpp:
```diff
@@ -0,0 +1,43 @@
+
+void populate_upgraders_map(const std::unordered_map<std::string, std::string>& content) {
+    // make sure we populate the map only once
+    UpgradersMap::lock->lock();
```
Use std::lock_guard

qihqi commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -0,0 +1,25 @@
+namespace jit {
+
+struct UpgradersMap {
+    static std::unordered_map<std::string, std::string> content;
```
just declare those as globals.

tugsbayasgalan replied:
Yeah sounds good.

qihqi replied:
I like Yanan's suggestion. To be concrete: declare content etc as non-static member fields of UpgradeMap, move functions into that class as methods; then declare `extern UpgradeMap gUpgraderMap;`  as global

qihqi commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -0,0 +1,25 @@
+
+struct UpgradersMap {
+    static std::unordered_map<std::string, std::string> content;
+    static std::mutex *lock;
```
any reason for pointer instead of just `std::mutex lock;`

gmagogsfm replied:
+1
Also, we should probably use `shared_mutex` so read will be cheaper:
https://en.cppreference.com/w/cpp/thread/shared_mutex

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

Commit:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

qihqi commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -8,18 +8,25 @@
-// this is used for testing, so copying is not a perf issue
 TORCH_API std::unordered_map<std::string, std::string> dump_upgraders_map();
 
+static UpgradersMap upgradersMap;
```
nit: move this to .cpp

qihqi commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -8,18 +8,25 @@
-    static std::unordered_map<std::string, std::string> content;
-    static std::mutex lock;
-    static bool isPopulated;
+class UpgradersMap {
```
nit: move this to .cpp

qihqi reviewed (approved):
Looks good overall

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

gmagogsfm commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -0,0 +1,32 @@
+ public:
+  void set_content(const std::unordered_map<std::string, std::string>& content);
+  int count();
+  std::unordered_map<std::string, std::string> get_content();
```
Can this return a const ref instead?

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/operator_upgraders/upgraders.h:
```diff
@@ -0,0 +1,32 @@
+
+ private:
+  std::unordered_map<std::string, std::string> content_;
+  std::mutex lock;
```
`shared_mutex` is more appropriate I think:
https://en.cppreference.com/w/cpp/thread/shared_mutex

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/upgraders.cpp:
```diff
@@ -0,0 +1,49 @@
+    return;
+  }
+
+  for (const auto& entry : content) {
+    content_.insert(entry);
+  }
```
We don't actually want to error out here because otherwise subsequent torch.jit.load would fail

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).

tugsbayasgalan committed:
Update on "Populate runtime with upgrader graph"

In this PR, we create a global empty map of upgrader graphs and lazily populate it with upgrader graph contents when the torch.jit.load is first invoked. We also add test APIs to add/remove test upgrader graphs to the upgrader_graph map. These APIs are all private. 


Differential Revision: [D32603258](https://our.internmc.facebook.com/intern/diff/D32603258)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32603258).
